### DISCUSSION TITLE
Best practices for ratatui apps

### DISCUSSION THREAD

**Discussion by kdheepak** [ratatui/ratatui#220] [Discussion]
Hi everyone. 

Thanks for maintaining and developing this framework. It is an awesome way to build TUI applications. I really like that there's a [starter template](https://github.com/tui-rs-revival/rust-tui-template/tree/3cdb6b0a94582beb05546cb1324eb0934d0de069) for beginners. 

I wanted to start a thread to discuss some additional best practices for apps, especially for apps that are slightly more complex than a single page TUI. 

I'm open to feedback on improving this examples. Feel free to add your own best practices too. And once we have collected a number of best practices, we can move them to the Wiki.

## Organize code with ratatui

As mentioned in the README, ratatui is a [immediate mode rendering application](https://www.youtube.com/watch?v=Z1qyvQsjK5Y). This necessitates a "main" loop that gets user events, updates app state, and renders the application.

For more complex applications it may make sense to organize code with standard GUI development principles.
Generally speaking, it seems to me that using a Model-View-Controller (MVC) pattern can be helpful for organizational purposes.

1) state / app, i.e. model,
2) ui, i.e. view,
3) main loop, i.e. controller.

```rust
struct Model {
    counter: i32,
    running: bool,
}

struct View;

impl View {
    fn render(&self, model: &Model, frame: &mut tui::Frame<tui::backend::CrosstermBackend<std::io::Stdout>>) {
      frame.render_widget(
          Paragraph::new(format!(
              "Counter: {}",
              app.counter
          ))
          .block(Block::default()),
          frame.size(),
      )
    }
}

struct Controller {
    model: Model,
    view: View,
    tui: Tui, // See https://github.com/tui-rs-revival/rust-tui-template/blob/3cdb6b0a94582beb05546cb1324eb0934d0de069/src/tui.rs
}

impl Controller {
    fn handle_key_events(&mut self, key_event: KeyEvent) -> Result<()> {
        match key_event {
            KeyCode::Char('+') => self.model.counter += 1,
            KeyCode::Char('-') => self.model.counter -= 1,
            KeyCode::Char('q') => self.model.running = false,
            _ => {},
        }
    }

    fn draw(&mut self, frame: &mut tui::Frame<tui::backend::CrosstermBackend<std::io::Stdout>>) -> Result<()> {
        self.view.draw(&self.model, &mut frame);
    }

    fn tick(&mut self) {
        std::thread::sleep(1);
    }

    fn run(&mut self) -> Result<()> {
        while self.model.running {
            match self.tui.events.next()? {
                Event::Tick => self.tick(),
                Event::Key(key_event) => self.handle_key_events(key_event)?,
            }
            terminal.draw(|f| self.draw(f))?;
        }
    }
}
```

## Useful error messages on `.unwrap()`

I personally like to add a hook for panics that 

1) calls `destruct_terminal()` before spitting out a stacktrace
2) uses [`better_panic`](https://github.com/mitsuhiko/better-panic) to give a stacktrace about what went wrong

```rust
fn destruct_terminal() {
  disable_raw_mode().unwrap();
  execute!(io::stdout(), LeaveAlternateScreen, DisableMouseCapture).unwrap();
  execute!(io::stdout(), cursor::Show).unwrap();
}

fn main() -> Result<()> {
  panic::set_hook(Box::new(|panic_info| {
    destruct_terminal();
    better_panic::Settings::auto().create_panic_handler()(panic_info);
  }));
  // ...
}
```

## Logging

There are a couple of ways to do this. 

1) You can design a logging widget that shows up only for debugging. 
2) You can generate log files that track the state of your applications. 

Depending on what you are debugging you may want one and/or the other.
In my application, I've used the latter approach. 

I use [`log4rs`](https://github.com/estk/log4rs) so that users can generate logs and share when debugging.
Here's an example implementation of [`initialize_logging()` in `taskwarrior-tui`](https://github.com/kdheepak/taskwarrior-tui/blob/c533f28a18509692ea5db1fd50b22aa92191b95b/src/main.rs#L60-L92).
After calling `initialize_logging()` in `main()`, I'm able to use `info!()` or `debug!()` anywhere in my application and get log files that show me what is going on while running the TUI.

Here's an example of using a logging widget in [`toot-rs`](https://github.com/toot-rs/toot-rs/blob/dbc31d2d752a2cbacaa6419ce90bc3b563f750a5/src/logging.rs).

## Configuration

You probably want to use [`directories` (high-level)](https://github.com/dirs-dev/directories-rs) or [`dirs` (low-level)](https://github.com/dirs-dev/dirs-rs) to manage configuration files using the `XDG` specification. This allows you to create configuration files that are stored where users might expect them to be in a cross platform manner.

In terms of the exact configuration file format, you can choose whatever you like. 

[`config-rs`](https://github.com/mehcode/config-rs) lets you read `.json`, `.toml`, `.ini` etc files by defining a `struct` in Rust that represents your configuration.

In [`xplr`](https://github.com/sayanarijit/xplr), configuration can be done using lua files. `xplr` embeds LuaJIT and allows for a sophisticated configuration system. This allows mapping keys in different layers, deciding what views should look like etc, all from the configuration file.

## Mapping key inputs

For simple TUI applications where you are not interested in adding configuration options, you can match and dispatch on hard-coded keycodes directly.

If you want to make it configurable however, you will have to use a configuration struct that stores key events mapped to a `Command`. 

A `Command` is a reified method call.
See the `Command` pattern in <http://gameprogrammingpatterns.com/command.html> for more information.

This is also typically called an `Action` or `Message`. 

```rust
impl Controller {
 fn handle_key_events(&mut self, key_event: KeyEvent) -> Result<()> {
    if key_event.code == c.keyconfig.quit {
        c.dispatch(Command::Quit);
    } else if key_event.code == app.keyconfig.increment_one {
        c.dispatch(Command::IncrementCounter(1));
    } else if key_event.code == app.keyconfig.decrement_one {
        c.dispatch(Command::DecrementCounter(1));
    }
    Ok(())
  }

  fn dispatch(&mut self, command: Command) {
    match command {
      Command::IncrementCounter(i) => self.model.counter += i,
      Command::DecrementCounter(i) => self.model.counter -= i,
      Action::Quit => self.model.running = false,
    }
  }
}
```

While you don't really _need_ to map `key_event`s to `Action`s, it can be useful to simplify logic and make the code more extensible and maintainable. 

For example, above we can easily add a larger `increment/decrement` by adding the following to the `handle_key_events` function:

```rust
    } else if key_event.code == app.keyconfig.increment_10 {
        c.dispatch(Command::IncrementCounter(10));
    } else if key_event.code == app.keyconfig.decrement_10 {
        c.dispatch(Command::DecrementCounter(10));
```

This kind of code organization can also help with testing and debugging.

**Comment by joshka** [comment]
Nice one.

I think I would generally advise new apps to choose https://crates.io/crates/tracing over log4rs, as it has a much bigger ecosystem (larger user base, 40-50x the daily downloads, huge amount of places to send the logs depending on the use case, really configurable)

There are some interesting crates that I've seen around using Actions / Commands. And there's a KeyBindings abstraction that can be useful there too. [ModalKit](https://crates.io/crates/modalkit) looks pretty interesting for that (though it's editor focused).

One thing I struggled to find a good answer to in [toot-rs](https://github.com/toot-rs/toot-rs) (especially as it was my first rust project) was how to communicate between views. I settled on an approach that effectively made the Model and the Controller one piece (this is similar to the Component approach used in two database TUIs: [GoBang](https://github.com/TaKO8Ki/gobang/tree/main/src/components) and [EDMA](https://github.com/nomadiz/edma/tree/master/gui/src/components). I didn't use a dispatch approach, but instead bubbled the events through to each component.

There's a number of different design patterns around that concept that could be expanded upon:
- How to handle the idea of focused component that accept events (text / mouse)
- communicating between components (e.g. channels / commands / methods)

One thing I did in toot-rs was to make several event sources all get handled in one place:

https://github.com/toot-rs/toot-rs/blob/dbc31d2d752a2cbacaa6419ce90bc3b563f750a5/src/event.rs#L16-L94
```rust
#[derive(Debug)]
pub struct Events {
    pub tx: Sender<Event>,
    rx: Receiver<Event>,
}

/// A wrapper around 3 event sources:
/// - Tick events
/// - Crossterm events
/// - Signals
impl Events {
    pub fn new() -> Self {
        let (tx, rx) = channel(100);
        Self { tx, rx }
    }

    pub fn start(&self) -> Result<()> {
        info!("Starting event loop");
        self.spawn_tick_task();
        self.spawn_signal_task()?;
        self.spawn_crossterm_task();
        Ok(())
    }

    /// Returns the next event
    /// This function is async and will block until an event is received
    /// or the channel is closed.
    /// If the channel is closed, `None` is returned.
    /// If the channel is not closed, `Some(Event)` is returned.
    pub async fn next(&mut self) -> Option<Event> {
        self.rx.recv().await
    }

    /// Sends a tick event every `tick_rate`
    fn spawn_tick_task(&self) -> JoinHandle<()> {
        let tx = self.tx.clone();
        let mut interval = interval(TICK_RATE);
        tokio::spawn(async move {
            loop {
                interval.tick().await;
                trace!("tick");
                if tx.send(Event::Tick).await.is_err() {
                    error!("Failed to send tick event");
                    break;
                }
            }
        })
    }

    /// Handle signals so killing the process cleans up the terminal correctly
    fn spawn_signal_task(&self) -> Result<JoinHandle<()>> {
        let tx = self.tx.clone();
        let mut signals = Signals::new([SIGHUP, SIGTERM, SIGINT, SIGQUIT])?;
        Ok(tokio::spawn(async move {
            while let Some(signal) = signals.next().await {
                match signal {
                    SIGTERM | SIGINT | SIGQUIT => {
                        info!("Received signal {}, shutting down", signal);
                        // Shutdown the system;
                        if tx.send(Event::Quit).await.is_err() {
                            error!("Failed to send quit event");
                            break;
                        }
                    }
                    _ => {
                        error!("Received unexpected signal {}", signal);
                    }
                }
            }
        }))
    }

    /// Sends crossterm events to the event channel
    fn spawn_crossterm_task(&self) -> JoinHandle<()> {
        let tx = self.tx.clone();
        let mut events = EventStream::new();
        tokio::spawn(async move {
            while let Some(Ok(event)) = events.next().await {
                trace!(crossterm_event = ?event);
                if tx.send(Event::Crossterm(event)).await.is_err() {
                    error!("Failed to send event to channel");
                    break;
                }
            }
        })
    }
}
```

One approach to building up some of this might be to start writing the rust book that @nickiel12 wants to do. We could do it in the main repo or in another repo in the tui-rs-revival org.

**Reply by kdheepak** [reply]
Thanks for sharing your event code @joshka !

Here's the same thing using Tokio's `select!` macro instead:

```rust
use std::time::Duration;

use anyhow::{anyhow, Context, Result};
use crossterm::event::{Event as CrosstermEvent, EventStream, KeyEvent, KeyEventKind, MouseEvent};
use futures::{FutureExt, StreamExt};
use signal_hook::consts::{SIGHUP, SIGINT, SIGQUIT, SIGTERM};
use signal_hook_tokio::{Signals, SignalsInfo};
use tokio::sync::mpsc;
use tracing::{error, info, trace};

#[derive(Clone, Copy, Debug)]
pub enum Event {
  Quit,
  Closed,
  Tick,
  Key(KeyEvent),
  Mouse(MouseEvent),
  Resize(u16, u16),
}

#[derive(Debug)]
pub struct EventHandler {
  sender: mpsc::UnboundedSender<Event>,
  receiver: mpsc::UnboundedReceiver<Event>,
}

impl EventHandler {
  pub fn new(tick_rate: u64) -> Self {
    let tick_rate = std::time::Duration::from_millis(tick_rate);

    let (sender, receiver) = mpsc::unbounded_channel();
    let _sender = sender.clone();

    tokio::spawn(async move {
      let mut reader = crossterm::event::EventStream::new();
      let mut signals =
        Signals::new([SIGHUP, SIGTERM, SIGINT, SIGQUIT]).context(anyhow!("Unable to create Signals handler.")).unwrap();
      let mut interval = tokio::time::interval(tick_rate);
      loop {
        let delay = interval.tick();
        let crossterm_event = reader.next().fuse();
        let signal = signals.next().fuse();

        tokio::select! {
          maybe_event = crossterm_event => {
            match maybe_event {
              Some(Ok(evt)) => {
                match evt {
                  CrosstermEvent::Key(key) => {
                    if key.kind == KeyEventKind::Press {
                      _sender.send(Event::Key(key)).unwrap();
                    }
                  },
                  CrosstermEvent::Resize(x, y) => {
                    _sender.send(Event::Resize(x, y)).unwrap();
                  },
                  _ => {},
                }
              }
              Some(Err(_)) => {}
              None => {},
            }
          },
          maybe_signal = signal => {
            match maybe_signal {
              Some(s) => {
                match s {
                    SIGTERM | SIGINT | SIGQUIT => {
                        info!("Received signal {}, shutting down", s);
                        _sender.send(Event::Quit).unwrap()
                    }
                    _ => {
                        error!("Received unexpected signal {}", s);
                    }
                }
              }
              None => {},
            }
          },
          _ = delay => {
              _sender.send(Event::Tick).unwrap();
          },
        }
      }
    });

    Self { sender, receiver }
  }

  pub async fn next(&mut self) -> Option<Event> {
    self.receiver.recv().await
  }
}
```

**Comment by orhun** [comment]
Thanks for putting this together! There are definitely some gems in there. I also wonder about your approach on testing tui apps as well! I remember some techniques you've used in [taskwarrior-tui](https://github.com/kdheepak/taskwarrior-tui). It would be great to document them.

One thing to add from my side about logging: I have used [`tui-logger`](https://github.com/gin66/tui-logger) widget in my [systeroid](https://github.com/orhun/systeroid) project and it works really well.

See https://github.com/orhun/systeroid/#logging

<details>
<summary>Demo</summary>

![demo](https://github.com/orhun/systeroid/raw/main/assets/systeroid-tui-logs.gif)

</details>

Another thing to note, there is [rust-tui-template](https://github.com/tui-rs-revival/rust-tui-template) available if anyone wants to bootstrap a tui project quickly.

**Reply by kdheepak** [reply]
Thanks for suggesting `tui-logger`! It's been great!

Here's my example code for creating a log that dumps to a file and sets up `tui-logger` using `tracing`:

```rust
use std::path::PathBuf;

use anyhow::{Context, Result};
use colored::Colorize;
use directories::ProjectDirs;
use env_logger;
use tracing_subscriber::{
  self, filter::EnvFilter, prelude::__tracing_subscriber_SubscriberExt, util::SubscriberInitExt, Layer,
};

pub fn initialize_logging() -> Result<()> {
  let directory = if let Ok(s) = std::env::var("MY_APP_NAME_HERE_DATA") {
    PathBuf::from(s)
  } else if let Some(proj_dirs) = ProjectDirs::from("com", "kdheepak", "my-app-name-here") {
    proj_dirs.data_local_dir().to_path_buf()
  } else {
    let s = "Error".red().bold();
    eprintln!("{s}: Unable to find data directory for my-app-name-here");
    std::process::exit(1)
  };
  std::fs::create_dir_all(directory.clone()).context(format!("{directory:?} could not be created"))?;
  let log_path = directory.join("my-app-name-here-debug.log");
  let log_file = std::fs::File::create(&log_path)?;
  let file_subscriber = tracing_subscriber::fmt::layer()
    .with_file(true)
    .with_line_number(true)
    .with_writer(log_file)
    .with_target(false)
    .with_ansi(false)
    .with_filter(EnvFilter::from_default_env());
  tracing_subscriber::registry().with(file_subscriber).with(tui_logger::tracing_subscriber_layer()).init();
  tui_logger::set_default_level(log::LevelFilter::Info);
  Ok(())
}
```

It puts the log file in a path that is found in either a environment variable or in a `XDG_*` directory.

**Comment by joshka** [comment]
We've spawned https://github.com/tui-rs-revival/ratatui-book from this and the related discord thread: https://discord.com/channels/1070692720437383208/1115053951000268832

**Comment by geofflittle** [comment]
Is there a best practice for sharing state across components in the component-based design?

**Reply by ProHaller** [reply]
I don't about best practices per se. But I found [this recipe](https://ratatui.rs/recipes/apps/terminal-and-event-handler/#_top) quite useful when building my game.

**Reply by joshka** [reply]
If the shared data is immutable, then just using a ref with a lifetime is good enough. The problems start when you're talking about shared immutable access to data.

https://doc.rust-lang.org/std/sync/index.html#higher-level-synchronization-objects has a pretty good list for multi-threaded shared synchronization of state.

Wrapping your shared data in a struct with an `Arc<Mutex<InnerData>>` field is often a easy default for apps. It allows you to derive clone on the struct, and write methods that appropriately lock and modify the inner data without leaking the details about locking / mutation outside the item. E.g.:

```rust
#[derive(Clone)]
struct State {
    inner: Arc<Mutex<InnerState>>,
}

struct InnerState {
    name: String,
    age: u8,
    // etc.
}

impl State {
    fn set_age(&self, age: u8) {
        let mut inner = self.inner.lock().unwrap();
        inner.age = age;
    }
}
```

You can similarly use `Rc<Cell>` or `Rc<RefCell>` for single-threaded apps. See https://doc.rust-lang.org/std/rc/index.html and https://doc.rust-lang.org/std/cell/index.html#introducing-mutability-inside-of-something-immutable

If your shared data is small, then using Atomics can be an approach that works (wrapped in an appropriate Arc as needed).

But each app is different, there's no real best practice, just practices that work for your situation. Some apps might be better to construct with a message passing approach. You'd choose this when the changes that happen to state are more important than the state itself. It's also useful when there's a large disconnect in the ownership hierarchy between two places which access similar data (e.g. a popup window and a status bar might share a distant parent in a layout, but not directly).

You might also consider implementing your app's items as Actors (there's a good pattern for tokio apps in https://ryhl.io/blog/actors-with-tokio/, but it can easily use non tokio channels just as easily).

Another general common pattern to use in general terms is to pass around a Context type object between the appropriate calls, which has all the data that the app needs to use. I think from memory yazi used to use this approach heavily, but it's changed a lot since I last read the source code. This just looks like having a parameter like `context: &mut AppContext` everywhere that's called and ensuring that you pass it down / around as needed. This probably gets complex though, but could be reasonable if you're building out library / framework type code.
